File: internal/session/title.go
package session

import (
	"strings"
	"unicode"

	"github.com/Dyu-36/gotack/internal/attachments"
	"github.com/Dyu-36/gotack/internal/engineapi"
)

func isDefaultTitle(title string) bool {
	switch strings.ToLower(strings.TrimSpace(title)) {
	case "", "new session", "new conversation", "new chat", "untitled session", "hội thoại mới":
		return true
	default:
		return false
	}
}

func titleFromMessages(messages []engineapi.Message) string {
	for _, message := range messages {
		if message.Role != "user" {
			continue
		}
		parts := engineapi.ExtractParts(message.Parts)
		text, refs := attachments.ParseAttachmentBlocks(parts.Text)
		if text == "" {
			var names []string
			for _, ref := range refs {
				names = append(names, attachments.BaseName(ref.FileName))
			}
			for _, attachment := range parts.Attachments {
				names = append(names, attachments.BaseName(attachment.FileName))
			}
			text = strings.Join(names, ", ")
		}
		text = strings.Map(func(r rune) rune {
			if unicode.IsControl(r) {
				return ' '
			}
			return r
		}, text)
		title := []rune(strings.Join(strings.Fields(text), " "))
		if len(title) == 0 {
			continue
		}
		if len(title) > 80 {
			return strings.TrimSpace(string(title[:79])) + "…"
		}
		return string(title)
	}
	return ""
}

File: frontend/src/features/conversations/title.ts
export const NEW_CONVERSATION_TITLE = 'Hội thoại mới'

const defaultTitles = new Set(['', 'new session', 'new conversation', 'new chat', 'untitled session', 'hội thoại mới'])

export function isDefaultTitle(title: string): boolean {
  return defaultTitles.has(title.trim().toLowerCase())
}

export function conversationTitle(title: string, text = '', fileNames: readonly string[] = []): string {
  if (!isDefaultTitle(title)) return title.trim()
  const clean = (text.trim() || fileNames.join(', ')).replace(/[\p{Cc}\s]+/gu, ' ').trim()
  const chars = Array.from(clean)
  return chars.length > 80 ? `${chars.slice(0, 79).join('').trimEnd()}…` : clean || NEW_CONVERSATION_TITLE
}
